File: utils/Step1_Firm_Generation.R
#-----------------------------------------------------------------------------------
#Step 1 Firm Synthesis
#-----------------------------------------------------------------------------------

# rm(list = ls())
#path2code <- '/Users/xiaodanxu/Documents/GitHub/SynthFirm/utils/'
# source('step0_SynthFirm_starter.R') # load packages
# # source('scenario/scenario_variables.R')  # load environmental variable
# source('config.R')
# path2file <-
#   "/Users/xiaodanxu/Documents/SynthFirm.nosync"
# setwd(path2file)
#
print("Start synthetic firm generation...")

# cbp <- data.table::fread(paste0(path2file, "/inputs_", scenario, '/', cbp_file), h = T)
# mzemp <-
#   data.table::fread(paste0(path2file, "/inputs_", scenario, '/', mzemp_file), h = T)
#
# c_n6_n6io_sctg <-
#   data.table::fread(paste0(path2file, "/SynthFirm_parameters/", c_n6_n6io_sctg_file), h = T)
# employment_per_firm <-
#   data.table::fread(paste0(path2file, "/SynthFirm_parameters/", employment_per_firm_file), h = T)
# employment_per_firm_gapfill <-
#   data.table::fread(paste0(path2file, "/SynthFirm_parameters/", employment_per_firm_gapfill_file), h = T)

output_dir <- paste0(path2file, "/outputs_", out_scenario)
if (!dir.exists(output_dir)){
  dir.create(output_dir)
}else{
  print("output directory exists")
}

################ part 1 ######################
#-----------------------------------------------------------------------------------
#Enumerating firms and merge with correspondenses
#-----------------------------------------------------------------------------------
print("Enumerating Firms")


#Aggregate the employment data by zones, NAICS, and firm size category
#1 = '1-19',2 = '20-99',3 ='100-499',4 = '500-999',5 = '1,000-2,499',6 = '2,500-4,999',7 = 'Over 5,000'

cbp <- cbp %>% mutate(employment = ifelse(employment < establishment, establishment, employment)) %>% as_tibble()
total_employment = sum(cbp$employment)

cbp <- as.data.table(cbp)
cbp_by_location <-
  cbp[!is.na(CBPZONE) &
        !is.na(FAFZONE) & !is.na(Industry_NAICS6_CBP),
      list(
        e1 = sum(e1),
        e2 = sum(e2),
        e3 = sum(e3),
        e4 = sum(e4),
        e5 = sum(e5),
        e6 = sum(e6),
        e7 = sum(e7)
      ),
      by = list(Industry_NAICS6_CBP, CBPZONE, FAFZONE)] #Remove records with missing zones and NAICS codes

setkey(c_n6_n6io_sctg, Industry_NAICS6_CBP)

cbp_by_industry <-
  merge(cbp_by_location, c_n6_n6io_sctg[, list(Industry_NAICS6_CBP, Industry_NAICS6_Make, Commodity_SCTG)], by =
          "Industry_NAICS6_CBP") #Merge in the I/O NAICS codes and SCTG codes

cbp_by_industry[, c("n2", "n4") := list(substr(Industry_NAICS6_CBP, 1, 2),
                                  substr(Industry_NAICS6_CBP, 1, 4))] #add 2 and 4 digit NAICS

cbp_long <-
  data.table::melt(
    cbp_by_industry,
    measure.vars = paste0("e", 1:7),
    variable.name = "esizecat",
    value.name = "est"
  ) #Melt to create separate rows for each firm size category

cbp_long <- as.data.table(cbp_long)
cbp_long[, esizecat := as.integer(esizecat)] #convert esizecat to an integer (1:7)
cbp_long <- cbp_long %>% filter(est > 0) %>% as_tibble()

# assign employment size per category
# cbp[, Emp := c(10L, 60L, 175L, 375L, 750L, 1750L, 3750L, 7500L)[esizecat]] # Estimate the number of employees
employment_per_firm <- employment_per_firm %>% select(NAICS, size_group, emp_per_est) %>% as_tibble()
cbp_long <- merge(cbp_long, employment_per_firm, by.x = c('Industry_NAICS6_CBP', 'esizecat'),
                  by.y = c('NAICS', 'size_group'), all.x = TRUE)

emp_gap_fill <- pull(employment_per_firm_gapfill, 'emp_per_est')
cbp_long <- cbp_long %>% mutate(emp_per_est = ifelse(is.na(emp_per_est), ifelse(esizecat ==1, emp_gap_fill[1],
                                                                         ifelse(esizecat ==2, emp_gap_fill[2], 
                                                                         ifelse(esizecat ==3, emp_gap_fill[3], 
                                                                         ifelse(esizecat ==4, emp_gap_fill[4], 
                                                                         ifelse(esizecat ==5, emp_gap_fill[5], 
                                                                         ifelse(esizecat ==6, emp_gap_fill[6], emp_gap_fill[7])))))), 
                                                     emp_per_est)) %>% as_tibble()

cbp_long <- as.data.table(cbp_long)
firms <-
  cbp_long[rep(seq_len(cbp_long[, .N]), est),] #Enumerates the agent businesses using the est variable.
firms[, BusID := .I] #Add an ID

emp_obs <- cbp %>%
  group_by(Industry_NAICS6_CBP, CBPZONE, FAFZONE) %>%
  summarise(total_emp_obs = sum(employment)) %>% as_tibble()

emp_sim <- firms %>%
  group_by(Industry_NAICS6_CBP, CBPZONE, FAFZONE) %>%
  summarise(total_emp_sim = sum(emp_per_est)) %>% as_tibble()

emp_adj <- emp_obs %>% left_join(emp_sim, by = c('Industry_NAICS6_CBP', 'CBPZONE', 'FAFZONE'))
emp_adj <- emp_adj %>% mutate(emp_adj = total_emp_obs/total_emp_sim)

firms <- firms %>% left_join(emp_adj, by = c('Industry_NAICS6_CBP', 'CBPZONE', 'FAFZONE')) %>% as_tibble()
firms <- firms %>% mutate(emp_per_est = emp_per_est * emp_adj)

# summarise(cbp_emp)
total_employment_est = sum(firms$emp_per_est)

print('total number of firm is ')
print(sum(cbp_long$est))
print('total number of employment is ')
print(total_employment_est)
#-----------------------------------------------------------------------------------
# Allocating specific commodity and location for each establishment
#-----------------------------------------------------------------------------------
# print("Allocating commodities and locations to establishments")



#Assign firms within study areas 
firms <- as.data.table(firms)
firms_in_boundary <- firms[CBPZONE > 999, list(CBPZONE, BusID, n2, emp_per_est)]
#Assign specific NAICS categories which would be used to locate businesses to tazs
firms_in_boundary[n2 %in% c("31", "32", "33"), n2 := "3133"]
firms_in_boundary[n2 %in% c("44", "45"), n2 := "4445"]
firms_in_boundary[n2 %in% c("48", "49"), n2 := "4849"]
firms_in_boundary[n2 %in% c("92"), n2 := "92"] 

emp_ranking_in_boundary <- data.table(melt(mzemp, id.vars = c("COUNTY", "MESOZONE")))
emp_ranking_in_boundary[, n2 := sub("rank", "", as.character(variable))]
setnames(emp_ranking_in_boundary, c("COUNTY", "value"), c("CBPZONE", "EmpRank"))
firms_in_boundary <-
  merge(firms_in_boundary,
        emp_ranking_in_boundary,
        c("CBPZONE", "n2"),
        allow.cartesian = TRUE,
        all.x = TRUE) # Merge the rankings dataset to the firms database based on county

#Select candidate tazs based on the industry of the firm, firm size, and ranking of that particular industry in a census block group
firms_in_boundary[, candidate := 0L]
firms_in_boundary[emp_per_est > 5000 & EmpRank %in% c(9, 10), candidate := 1L]
firms_in_boundary[emp_per_est > 2000 &
                    emp_per_est <= 5000 & EmpRank %in% c(7:10), candidate := 1L]
firms_in_boundary[emp_per_est > 500 &
                    emp_per_est <= 2000 & EmpRank %in% c(5:10), candidate := 1L]
firms_in_boundary[emp_per_est > 100 & emp_per_est <= 500 & EmpRank %in% c(4:10), candidate := 1L]
firms_in_boundary[emp_per_est > 20 & emp_per_est <= 100 & EmpRank %in% c(2:10), candidate := 1L]
firms_in_boundary[emp_per_est <= 20 & EmpRank %in% c(1:10), candidate := 1L]
#small number of businesses that did not get a candiate TAZ - allow those to have some candidates (small error is better than omitting the businesses)
ZeroCand <- firms_in_boundary[, sum(candidate), by = BusID][V1 == 0]
firms_in_boundary[BusID %in% ZeroCand$BusID, candidate := 1L]
firms_in_boundary <- firms_in_boundary[candidate == 1,] #remove non-candidate TAZs
firms_in_boundary[, u := runif(.N)] #Generate a random number based on which one of the candidate tazs would be selected
firms_in_boundary <-
  firms_in_boundary[firms_in_boundary[, .I[which.max(u)], by = BusID]$V1,] #Assign the taz for which the random number generated is the highest among all candidate tazs

#Assign MESOZONES for all firms
firms[CBPZONE <= 999, MESOZONE := CBPZONE + 20000L]
#cbp[CBPZONE %in% 801:808, MESOZONE:= CBPZONE]
setkey(firms, BusID)
setkey(firms_in_boundary, BusID)
firms[CBPZONE > 999, MESOZONE := firms_in_boundary$MESOZONE]
#Cleanup
rm(mzemp, ZeroCand, firms_in_boundary)
firms[, c("Industry_NAICS6_CBP", "total_emp_obs", "total_emp_sim", "emp_adj", "n2", "n4", "est") := NULL] #Revome extra fields,
setnames(firms, "emp_per_est", "Emp")
write.csv(firms, paste0(output_dir, '/', synthetic_firms_no_location_file), row.names=FALSE)
print('synthetic firm generation is done!')

File: scenario_analysis/national_input_generation/Step1_Firm_Generation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Jan 16 10:29:16 2024

@author: xiaodanxu
"""

import pandas as pd
import os
import numpy as np
from pandas import read_csv



########################################################
#### step 1 - configure environment and load inputs ####
########################################################

scenario_name = 'Seattle'
out_scenario_name = 'Seattle'
file_path = '/Users/xiaodanxu/Documents/SynthFirm.nosync'
parameter_dir = 'SynthFirm_parameters'
number_of_processes = 4
input_dir = 'inputs_' + scenario_name
output_path = 'outputs_' + out_scenario_name

os.chdir(file_path)    
cbp_file = os.path.join(input_dir, 'data_emp_cbp_imputed.csv')
mzemp_file = os.path.join(input_dir, 'data_mesozone_emprankings.csv')
mesozone_to_faf_file = os.path.join(input_dir, 'zonal_id_lookup_final.csv')

c_n6_n6io_sctg_file = os.path.join(parameter_dir, 'corresp_naics6_n6io_sctg_revised.csv')
employment_per_firm_file = os.path.join(parameter_dir, 'employment_by_firm_size_naics.csv')
employment_per_firm_gapfill_file = os.path.join(parameter_dir, 'employment_by_firm_size_gapfill.csv')
zip_to_tract_file = os.path.join(parameter_dir, 'ZIP_TRACT_LOOKUP_2016.csv')
synthetic_firms_no_location_file = os.path.join(output_path, 'synthetic_firms.csv')

# def synthetic_firm_generation(cbp_file, mzemp_file, c_n6_n6io_sctg_file, employment_per_firm_file,
#                               employment_per_firm_gapfill_file, synthetic_firms_no_location_file,
#                               output_path):
print("Start synthetic firm generation...")
# load model inputs
cbp = read_csv(cbp_file)
mzemp = read_csv(mzemp_file)
mesozone_to_faf = read_csv(mesozone_to_faf_file)
# mzemp = mzemp.drop(columns = ['cnty_id'])
c_n6_n6io_sctg = read_csv(c_n6_n6io_sctg_file)
employment_per_firm = read_csv(employment_per_firm_file)
employment_per_firm_gapfill = read_csv(employment_per_firm_gapfill_file)
zip_to_tract_crosswalk = read_csv(zip_to_tract_file)

# create result directory if not exist

if not os.path.exists(output_path):
    os.mkdir(output_path)
else:
  print("output directory exists")
    
    
    # <codecell>
    
    ########################################################
    #### step 2 - Enumerate list of firms and workers ######
    ########################################################
    
print("Enumerating Firms")
criteria = (cbp.loc[:, 'employment'] < cbp.loc[:, 'establishment'])
cbp.loc[criteria, 'employment'] = cbp.loc[criteria, 'establishment']


# drop invalid record (if any)
cbp = cbp.dropna(subset=['Industry_NAICS6_CBP', 'FAFZONE', 'CBPZONE'])

cbp_by_industry = pd.merge(cbp, c_n6_n6io_sctg, 
                            on = 'Industry_NAICS6_CBP', 
                            how = 'left')

cbp_by_industry.loc[:, 'n2'] = \
    cbp_by_industry.loc[:, 'Industry_NAICS6_CBP'].astype(str).str[0:2]
cbp_by_industry.loc[:, 'n4'] = \
    cbp_by_industry.loc[:, 'Industry_NAICS6_CBP'].astype(str).str[0:4]

cbp_long = pd.melt(cbp_by_industry, 
                    id_vars=["Industry_NAICS6_CBP", "CBPZONE", "FAFZONE", "COUNTY","ZIPCODE",
                             "Industry_NAICS6_Make", "Commodity_SCTG", "n2", "n4"], 
                    value_vars= ['e1', 'e2', 'e3', 'e4', 'e5', 'e6', 'e7'],
                    var_name='esizecat', value_name='est')

cbp_long.loc[:, 'esizecat'] = cbp_long.loc[:, 'esizecat'].str[1:2].astype(int)
cbp_long = cbp_long.loc[cbp_long['est'] > 0]

employment_per_firm_short = employment_per_firm[['NAICS', 'size_group', 'emp_per_est']]
cbp_long = pd.merge(cbp_long, employment_per_firm_short, 
                    left_on = ['Industry_NAICS6_CBP', 'esizecat'],
                    right_on = ['NAICS', 'size_group'], how = 'left')

cbp_long_to_fill = cbp_long.loc[cbp_long['emp_per_est'].isna()]
cbp_long_no_fill = cbp_long.loc[~cbp_long['emp_per_est'].isna()]

employment_per_firm_gapfill = \
    employment_per_firm_gapfill[['size_group', 'emp_per_est']]
cbp_long_to_fill = cbp_long_to_fill.drop(columns = ['size_group', 'emp_per_est'])
cbp_long_to_fill = pd.merge(cbp_long_to_fill, employment_per_firm_gapfill,
                            left_on = 'esizecat', right_on = 'size_group',
                            how = 'left')

cbp_long = pd.concat([cbp_long_no_fill, cbp_long_to_fill])

firms = pd.DataFrame(cbp_long.values.repeat(cbp_long.est, axis=0), 
                      columns=cbp_long.columns)

firms.loc[:, 'BusID'] = firms.reset_index().index + 1
# 8,396,679 firms nationwide

# <codecell>

# adjust employment

emp_obs = \
    cbp.groupby(['Industry_NAICS6_CBP','FAFZONE', 'CBPZONE'])[['employment']].sum() 
emp_obs.columns = ['emp_obs']
emp_obs = emp_obs.reset_index()

emp_sim = \
    firms.groupby(['Industry_NAICS6_CBP','FAFZONE', 'CBPZONE'])[['emp_per_est']].sum()
emp_sim.columns = ['emp_sim']
emp_sim = emp_sim.reset_index()

emp_adj = pd.merge( emp_obs, emp_sim, 
                    on = ['Industry_NAICS6_CBP', 'CBPZONE', 'FAFZONE'],
                    how = 'left')

emp_adj.loc[:, 'emp_adj'] = \
    emp_adj.loc[:, 'emp_obs'] / emp_adj.loc[:, 'emp_sim']
    
emp_adj = emp_adj[['Industry_NAICS6_CBP', 'CBPZONE', 'FAFZONE', 'emp_adj']]

firms = pd.merge(firms, emp_adj, 
                  on = ['Industry_NAICS6_CBP', 'CBPZONE', 'FAFZONE'],
                  how = 'left')
firms.loc[:, 'emp_per_est'] *= firms.loc[:, 'emp_adj']
# firms.loc[:, 'emp_per_est'] =np.round(firms.loc[:, 'emp_per_est'].astype(float), 2)

# validate employment
total_employment_est = firms.loc[:, 'emp_per_est'].sum()

print('total number of input firm is ')
print(cbp_long.loc[:, 'est'].sum())
print('total number of output firm is ')
print(len(firms))

print('total number of input employment is ')
print(cbp.loc[:, 'employment'].sum())
print('total number of output employment is ')
print(total_employment_est)

# <codecell>

########################################################################
# Step 3 - Allocating commodity and location for each establishment ####
########################################################################


#### scale employment to align with LEHD total (by NAICS 2-DIGIT code)
# pre-process employment ranking
mesozone_to_cbpzone = mesozone_to_faf[['MESOZONE', 'CBPZONE']]
lehd_emp_for_scaling = pd.merge(mzemp, mesozone_to_cbpzone, on = 'MESOZONE')

emp_colnames = ["rank11",
  "rank21",
  "rank22",
  "rank23",
  "rank3133",
  "rank42",
  "rank4445",
  "rank4849",
  "rank51",
  "rank52",
  "rank53",
  "rank54",
  "rank55",
  "rank56",
  "rank61",
  "rank62",
  "rank71",
  "rank72",
  "rank81",
  "rank92"]

lehd_emp_for_scaling.loc[:, emp_colnames] = lehd_emp_for_scaling.loc[:, emp_colnames].fillna(0)
lehd_emp_for_scaling = lehd_emp_for_scaling.groupby('CBPZONE')[emp_colnames].sum()
lehd_emp_for_scaling = lehd_emp_for_scaling.reset_index()
lehd_emp_for_scaling = pd.melt(lehd_emp_for_scaling, id_vars = ["CBPZONE"],
                               value_vars = emp_colnames,
                                  var_name= 'industry', value_name='emp_lehd')

lehd_emp_for_scaling = lehd_emp_for_scaling.reset_index()
lehd_emp_for_scaling.loc[:, 'industry'] = \
    lehd_emp_for_scaling.loc[:, 'industry'].str.split('rank').str[1]
    

print('total LEHD employment:')
print(lehd_emp_for_scaling.emp_lehd.sum())
# develop firm emp scaling factor
firms.loc[:, 'industry'] = firms.loc[:, 'n2']
firms.loc[firms['industry'].isin(["31", "32", "33"]), 'industry'] = "3133"
firms.loc[firms['industry'].isin(["44", "45"]), 'industry'] = "4445"
firms.loc[firms['industry'].isin(["48", "49"]), 'industry'] = "4849"

emp_sim = \
    firms.groupby(['industry', 'CBPZONE'])[['emp_per_est']].sum()
emp_sim.columns = ['emp_sim']
emp_sim = emp_sim.reset_index()

emp_adj = pd.merge( lehd_emp_for_scaling, emp_sim, 
                    on = ['industry', 'CBPZONE'],
                    how = 'left')

emp_adj.loc[:, 'emp_adj'] = \
    emp_adj.loc[:, 'emp_lehd'] / emp_adj.loc[:, 'emp_sim']
# emp_adj = emp_adj.fillna(0)
# if CBP < LEHD, do not adjust employment -> to keep employments from imputation step
emp_adj.loc[emp_adj['emp_adj']<1, 'emp_adj'] = 1

emp_adj = emp_adj[['industry', 'CBPZONE', 'emp_adj']]
firms.drop(columns = ['emp_adj'], inplace = True)
firms = pd.merge(firms, emp_adj, 
                  on = ['industry', 'CBPZONE'],
                  how = 'left')

firms.loc[:, 'emp_adj'].fillna(1, inplace = True)
firms.loc[:, 'emp_per_est'] *= firms.loc[:, 'emp_adj']
# firms.loc[:, 'emp_per_est'] =np.round(firms.loc[:, 'emp_per_est'].astype(float), 2)

# validate employment
total_employment_est = firms.loc[:, 'emp_per_est'].sum()

print('total number of output employment is ')
print(total_employment_est)

# <codecell>

# separate firms inside/outside study reion
emp_ranking_in_boundary = mzemp.dropna(subset = ['COUNTY'])
emp_ranking_in_boundary = \
    emp_ranking_in_boundary.rename(columns = {"COUNTY": 'CBPZONE'})
cbpzone_in_region = emp_ranking_in_boundary.CBPZONE.unique()


essential_attr = ['CBPZONE', 'FAFZONE',	'esizecat', 'Industry_NAICS6_Make', 'COUNTY', 'ZIPCODE',
                'Commodity_SCTG', 'emp_per_est', 'BusID', 'industry']
firms_out_boundary = \
    firms.loc[~firms['CBPZONE'].isin(cbpzone_in_region), essential_attr]

firms_in_boundary = \
    firms.loc[firms['CBPZONE'].isin(cbpzone_in_region), essential_attr]


print('number of firms outside study area:')
print(len(firms_out_boundary))

print('number of firms within study area:')
print(len(firms_in_boundary))

# <codecell>

# generate mesozone ID for firms outside boundary
firms_out_boundary = pd.merge(firms_out_boundary,
                              mesozone_to_cbpzone,
                              on = 'CBPZONE', how = 'left')

firms_out_boundary = \
    firms_out_boundary.groupby(essential_attr).sample(1, replace = False, random_state = 1)

# assign mesozone for firms within boundary

# separate firms with/without zipcode
firms_in_boundary_nozip = firms_in_boundary.loc[firms_in_boundary['ZIPCODE'] == 99999]
firms_in_boundary_withzip = firms_in_boundary.loc[firms_in_boundary['ZIPCODE'] != 99999]

# create tract ID from CBG id
emp_ranking_in_boundary.loc[:, 'MESOZONE'] = \
    emp_ranking_in_boundary.loc[:, 'MESOZONE'].astype(int).astype(str).str.zfill(12)
emp_ranking_in_boundary.loc[:, 'geoid'] = \
    emp_ranking_in_boundary.loc[:, 'MESOZONE'].str[0:11]

# <codecell>
zip_to_tract_crosswalk.loc[:, 'geoid'] = \
    zip_to_tract_crosswalk.loc[:, 'geoid'].astype(int).astype(str).str.zfill(11)
    
emp_ranking_in_boundary = pd.melt(emp_ranking_in_boundary, 
                                  id_vars = ['CBPZONE', 'MESOZONE', 'geoid'],
                                  value_vars = emp_colnames,
                                  var_name= 'industry', value_name='emp_lehd')

emp_ranking_in_boundary = emp_ranking_in_boundary.reset_index()
emp_ranking_in_boundary = \
    emp_ranking_in_boundary.loc[emp_ranking_in_boundary['emp_lehd']>0]
# if a CBG has 0 employment for selected industry, it is not a valid candidate
# therefore, CBGs with missing ranking is dropped
emp_ranking_in_boundary.loc[:, 'industry'] = \
    emp_ranking_in_boundary.loc[:, 'industry'].str.split('rank').str[1]
    
print('total LEHD employment within study area:')
print(emp_ranking_in_boundary.emp_lehd.sum())

# <codecell>

# assign mesozone to firms with zip code ID

industries = emp_ranking_in_boundary.loc[:, 'industry'].unique()
zip_to_tract_crosswalk.rename(columns = {'zip': 'ZIPCODE'}, inplace = True)
zip_to_tract_crosswalk.drop(columns = ['bus_ratio', 'fraction'], inplace = True)
print(len(firms_in_boundary_withzip))

# <codecell>
firms_out_withzip = None

for ind in industries:

    firms_to_assign = firms_in_boundary_withzip.loc[firms_in_boundary_withzip['industry'] == ind]
    print('numbers of firms to assign from industry = ' + str(ind))
    print(len(firms_to_assign.BusID.unique()))
    
    firms_to_assign = pd.merge(firms_to_assign, zip_to_tract_crosswalk,
                               on = 'ZIPCODE', how = 'left')
    firms_to_assign = pd.merge(firms_to_assign, emp_ranking_in_boundary,
                               on = ['CBPZONE', 'geoid', 'industry'], how = 'left')
    firms_to_assign['emp_lehd'] = firms_to_assign['emp_lehd'].fillna(1)
    # print(len(firms_to_assign.BusID.unique()))
    
    # Sometimes, LODES report 0 employment in a county, while firm data as non-zero
    # may attributed to imputation for non-payroll workers
    # fill minimum ranking for all zones as no information is available for the ranking
    
    firms_to_assign = \
        firms_to_assign.groupby(essential_attr).sample(1,
                                                        weights = firms_to_assign['emp_lehd'],
                                                        replace = False, random_state = 1)
    firms_to_assign = \
        firms_to_assign.drop(columns = ['index', 'industry', 'emp_lehd'])
    firms_to_assign.loc[:, 'MESOZONE'].fillna(method = 'ffill', inplace = True)
    firms_to_assign.loc[:, 'MESOZONE'].fillna(method = 'bfill', inplace = True)
    
    firms_out_withzip = pd.concat([firms_out_withzip, firms_to_assign])
    
    # break
print(len(firms_out_withzip))

# <codecell>

# assign mesozone to firms without zip code
firms_out_nozip = None

def split_dataframe(df, chunk_size = 10 ** 5): 
    chunks = list()
    num_chunks = len(df) // chunk_size + 1
    for i in range(num_chunks):
        chunks.append(df[i*chunk_size:(i+1)*chunk_size])
    return chunks

for ind in industries:

    firms_to_assign = firms_in_boundary_nozip.loc[firms_in_boundary_nozip['industry'] == ind]
    print('numbers of firms to assign from industry = ' + str(ind))
    print(len(firms_to_assign.BusID.unique()))
    chunks = split_dataframe(firms_to_assign, 10000)
    post_firms_to_assign = None
    for chunk in chunks:

        chunk = pd.merge(chunk, emp_ranking_in_boundary,
                                   on = ['CBPZONE', 'industry'], how = 'left')
        # print(len(firms_to_assign.BusID.unique()))
        chunk['emp_lehd'] = chunk['emp_lehd'].fillna(1)
        # print(len(firms_to_assign.BusID.unique()))
        
        chunk = \
            chunk.groupby(essential_attr).sample(1,
                                                 weights = chunk['emp_lehd'],
                                                 replace = False, random_state = 1)
        chunk = \
            chunk.drop(columns = ['index', 'industry', 'emp_lehd'])
        chunk.loc[:, 'MESOZONE'].fillna(method = 'ffill', inplace = True)
        chunk.loc[:, 'MESOZONE'].fillna(method = 'bfill', inplace = True)
        post_firms_to_assign = pd.concat([post_firms_to_assign, chunk])
    firms_out_nozip = pd.concat([firms_out_nozip, post_firms_to_assign])
    
    # break
print(len(firms_out_nozip))


# <codecell> 
# unique_counties = firms_in_boundary.CBPZONE.unique()
# firms_in_boundary_out = None
# # randomly assign CBG
# # loop through counties, so it doesn't explode memory usage

# for ct in unique_counties:
#     # print('assigning CBG within county fips ' + str(ct))
#     firms_in_boundary_sel = firms_in_boundary.loc[firms_in_boundary['CBPZONE'] == ct]
#     firms_in_boundary_sel = pd.merge(firms_in_boundary_sel, 
#                               emp_ranking_in_boundary,
#                               on = ["CBPZONE", "n2"],
#                               how = 'left') # Merge the rankings dataset to the firms database based on county
#     firms_in_boundary_sel['EmpRank'] = firms_in_boundary_sel['EmpRank'].fillna(1)
    

#     firms_in_boundary_sel = \
#         firms_in_boundary_sel.groupby(essential_attr).sample(1, 
#                                                         weights = firms_in_boundary_sel['EmpRank'],
#                                                         replace = False, random_state = 1)
#     firms_in_boundary_sel = \
#         firms_in_boundary_sel.drop(columns = ['index', 'industry', 'EmpRank'])
#     firms_in_boundary_sel.loc[:, 'MESOZONE'].fillna(method = 'ffill', inplace = True)
#     firms_in_boundary_sel.loc[:, 'MESOZONE'].fillna(method = 'bfill', inplace = True)
#     # for firms with no emp ranking, forward fill it with locations from nearest industry
#     # check na
#     firms_with_na = firms_in_boundary_sel.loc[firms_in_boundary_sel['MESOZONE'].isna()]
#     if len(firms_with_na) > 0:
#         print(str(len(firms_with_na)) + ' firms failed to have zone assigned')
#         print('and their firm ids are:')
#         print(firms_with_na.BusID.unique())
#         print('WARNING!!: The following pipeline can fail due to this error')
#         # break
#     firms_in_boundary_out = pd.concat([firms_in_boundary_out, firms_in_boundary_sel])
#     # break

    
# <codecell>

####################################################
# Step 4 - final formatting and writing outputs ####
####################################################

firms = pd.concat([firms_out_boundary, firms_out_withzip, firms_out_nozip])

print('number of firms before writing output:')
print(len(firms))
    
output_attr = ['CBPZONE', 'FAFZONE',	'esizecat', 'Industry_NAICS6_Make',
                'Commodity_SCTG', 'emp_per_est', 'BusID', 'MESOZONE', 'ZIPCODE']

firms = firms[output_attr]
firms = firms.rename(columns = {'emp_per_est': 'Emp'})
firms.to_csv(synthetic_firms_no_location_file, index = False)
